src/servers/http_server.cc

```diff
@@ -2868,11 +2772,173 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
+  {
+  }
+
+ private:
+  void Handle(evhtp_request_t* req) override;
+
+  re2::RE2 server_regex_;
+  re2::RE2 model_regex_;
+  re2::RE2 modelcontrol_regex_;
+  re2::RE2 systemsharedmemory_regex_;
+  re2::RE2 cudasharedmemory_regex_;
+};
+
+void
+HTTPAPIServer::Handle(evhtp_request_t* req)
+{
+  LOG_VERBOSE(1) << "HTTP request: " << req->method << " "
+                 << req->uri->path->full;
+
+  if (std::string(req->uri->path->full) == "/v2/models/stats") {
+    // model statistics
+    HandleModelStats(req);
+    return;
+  }
+
+  std::string model_name, version, kind;
+  if (RE2::FullMatch(
+          std::string(req->uri->path->full), model_regex_, &model_name,
+          &version, &kind)) {
+    if (kind == "ready") {
+      // model ready
+      HandleModelReady(req, model_name, version);
+      return;
+    } else if (kind == "infer") {
+      // model infer
+      HandleInfer(req, model_name, version);
+      return;
+    } else if (kind == "config") {
+      // model configuration
+      HandleModelConfig(req, model_name, version);
+      return;
+    } else if (kind == "stats") {
+      // model statistics
+      HandleModelStats(req, model_name, version);
+      return;
+    } else if (kind == "") {
+      // model metadata
+      HandleModelMetadata(req, model_name, version);
+      return;
+    }
+  }
+
+  std::string region, action, rest, repo_name;
+  if (std::string(req->uri->path->full) == "/v2") {
+    // server metadata
+    HandleServerMetadata(req);
+    return;
+  } else if (RE2::FullMatch(
+                 std::string(req->uri->path->full), server_regex_, &rest)) {
+    // server health
+    HandleServerHealth(req, rest);
+    return;
+  } else if (RE2::FullMatch(
+                 std::string(req->uri->path->full), systemsharedmemory_regex_,
+                 &region, &action)) {
+    // system shared memory
+    HandleSystemSharedMemory(req, region, action);
+    return;
+  } else if (RE2::FullMatch(
+                 std::string(req->uri->path->full), cudasharedmemory_regex_,
+                 &region, &action)) {
+    // cuda shared memory
+    HandleCudaSharedMemory(req, region, action);
+    return;
+  } else if (RE2::FullMatch(
+                 std::string(req->uri->path->full), modelcontrol_regex_,
+                 &repo_name, &kind, &model_name, &action)) {
+    // model repository
+    if (kind == "index") {
+      HandleRepositoryIndex(req, repo_name);
+      return;
+    } else if (kind.find("models", 0) == 0) {
+      HandleRepositoryControl(req, repo_name, model_name, action);
+      return;
+    }
+  }
+
+  LOG_VERBOSE(1) << "HTTP error: " << req->method << " " << req->uri->path->full
+                 << " - " << static_cast<int>(EVHTP_RES_BADREQ);
+
+  evhtp_send_reply(req, EVHTP_RES_BADREQ);
+}
+
+
+// Handle Sagemaker HTTP requests to inference server APIs
+class SagemakerAPIServer : public BaseAPIServer {
+ public:
```

Comment:
I would advice placing SagemakeAPIServer class in a file of its own.. 
call it `sagemaker_server.cc`.

Reply:
I agree, the `http_server.cc` seems to be able to be split into files for different endpoints (HTTP / metrics / sagemaker). But I think I will keep `http_server.h` as it is so conceptually it lists the endpoints that uses HTTP as the underlying network protocol. The alternative is also separating `http_server.h` into `xxx_server.h`, then the endpoints are represented separately, which I think is better from the encapsulation perspective.